google/cloud/storage/internal/range_from_pagination.h

```diff
@@ -0,0 +1,180 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_RANGE_FROM_PAGINATION_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_RANGE_FROM_PAGINATION_H_
+
+#include "google/cloud/status_or.h"
+#include "google/cloud/storage/version.h"
+#include <iterator>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+template <typename T, typename Range>
+class PaginationIterator {
+ public:
+  //@{
+  /// @name Iterator traits
+  using iterator_category = std::input_iterator_tag;
+  using value_type = StatusOr<T>;
+  using difference_type = std::ptrdiff_t;
+  using pointer = value_type*;
+  using reference = value_type&;
+  //@}
+
+  PaginationIterator() : owner_(nullptr) {}
+
+  PaginationIterator& operator++() {
+    *this = owner_->GetNext();
+    return *this;
+  }
+
+  PaginationIterator const operator++(int) {
+    PaginationIterator tmp(*this);
+    operator++();
+    return tmp;
+  }
+
+  value_type const* operator->() const { return &value_; }
+  value_type* operator->() { return &value_; }
+
+  value_type const& operator*() const& { return value_; }
+  value_type& operator*() & { return value_; }
+#if GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
+  value_type const&& operator*() const&& { return std::move(value_); }
+#endif  // GOOGLE_CLOUD_CPP_HAVE_CONST_REF_REF
+  value_type&& operator*() && { return std::move(value_); }
+
+  friend bool operator==(PaginationIterator const& lhs,
+                         PaginationIterator const& rhs) {
+    // All end iterators are equal.
+    if (lhs.owner_ == nullptr) {
+      return rhs.owner_ == nullptr;
+    }
+    // Iterators on different streams are always different.
+    if (lhs.owner_ != rhs.owner_) {
+      return false;
+    }
+    // Iterators on the same stream are equal if they point to the same object.
+    if (lhs.value_.ok() && rhs.value_.ok()) {
+      return lhs.value_.value() == rhs.value_.value();
+    }
+    return lhs.value_.status() == rhs.value_.status();
```

Comment:
Opt: I usually try to avoid `x.status()` when `x.ok()`.  Neither the (pre)Abseil documentation, nor ours, seem to guarantee that you get anything in particular.

Reply:
Good point, let me know if the change I made makes sense.